Implementation of the ELF format dumper

i-pankrat opened the pull request:
An alternative way to create an application checkpoint. The approach allows creating checkpoints without any additional privileges, in particular the root user privileges.

For the checkpoint, a similar ELF core file is generated, which was generated earlier when using the kernel's ability to create a core dump on the abort signal.

AntonKozlov reviewed (commented):
It will be nice to disable debugging logging by default (and in case that is still can be required -- to be able to turn that on by e.g. a flag or env variable)

i-pankrat commented:
It's not always possible to properly create an ELF core file for the restoration, still working on the issue. So pull request is draft now.

rvansa commented on minicriu-client.c:
```diff
@@ -149,40 +486,46 @@ int minicriu_dump(void) {
 
-	acts[MC_THREAD_SIG] = oldhnd;
+	// Say to other threads that barrier is initialized
+	__atomic_fetch_add(&mc_barrier_initialization, 1, __ATOMIC_SEQ_CST);
```
This is not really an elegant solution, you're using two primitives instead of one. If you don't want to count the number of threads ahead, maybe use a semaphore?
Anyway, this piece of code will be eventually rewritten to support more than 64 threads, and catch races between creating new threads and listing them above.

i-pankrat replied:
You have a point. I'll look into it on my weekend.

rvansa reviewed (commented):
I don't know what is the bar for integration, but I think this is what makes `minicriu` actually useful. I am all for integration and taking this as the starting point.

omikhaltsova commented on minicriu-client.c:
```diff
@@ -102,12 +144,295 @@ static int writefile(const char *file, const char *buf, size_t len) {
+				if (leftData) {
+					if (fwrite(paddingData, leftData, 1, coreFile)) {
+						perror("Failed replace map content with zeroes. Failed to create checkpoint.");
+						fclose(proc_maps);
```
I guess you want to close `coreFile` here, right?
Maybe it'd be better to refactor this func into smaller blocks? It helps to avoid such mistakes.

i-pankrat replied:
It really may be useful to split the function into two parts: one for reading data from `/proc/self/maps` and another to create `minicriu_core`.

I'll get into it this weekend.

omikhaltsova commented on minicriu-client.c:
```diff
@@ -102,12 +144,295 @@ static int writefile(const char *file, const char *buf, size_t len) {
+
+	char filename[32];
+	sprintf(filename, "minicriu-core.%d", pid);
+	FILE *coreFile = fopen(filename, "w+");
```
It's better to add a check that `coreFile` opened/created successfully.

i-pankrat replied:
Fix in [1cf34b2](https://github.com/CRaC/minicriu/pull/9/commits/1cf34b287c2d9ca113eeee0a7925f7b388c124db)

i-pankrat commented on minicriu-client.c:
```diff
@@ -102,12 +144,295 @@ static int writefile(const char *file, const char *buf, size_t len) {
+		}
+
+		// [vsyscall] is mapped to the same address in each process
+		if (!strncmp(buffer + name_start, "[vsyscall]", sizeof("[vsyscall]") - 1)) {
```
Sorry for such a long reply, I was busy with my studies. To tell you the truth, I'm not sure if `[vvar]` and `[vdso]` should be done the same way.  I don't save `[vsyscall]` for two reasons. First, the process has no rights to read the region and second, in every process the `[vsyscall]` region is mapped to the same address space. As far as I know, the same cannot be said for `[vvar]` and `[vsdo]`.

Moreover, new regions `[vvar]` and `[vdso]` are deleted during checkpoint recovery and old ones are restored in #4. This did not cause any problems on a simple example.

This is an important question, and I am investigating it.

Link to test example -- https://github.com/i-pankrat/crac/commit/a85581f5d0067f3bde07cb1c359a943f3b29d902?diff=unified

rvansa replied:
Sure, that aspect can be handled elsewhere.

In my case I was getting SIGSEGV (I think) when the code tried to read from `[vvar]`, despite that region being mapped with read privileges. And once the JVM is restored, it would try to read current timestamps; accessing the old addresses (in new process where vvar and vdso are on a different memory location) would cause another segfault. Luckily, `mremap` from the new addresses to the old ones was sufficient to fix this.

i-pankrat commented on minicriu-client.c:
```diff
@@ -42,13 +42,55 @@
+Elf64_Nhdr nhdr[MC_MAX_THREADS];
+struct elf_prstatus prstatus[MC_MAX_THREADS];
+
+static pthread_mutex_t mc_getregs_mutex = PTHREAD_MUTEX_INITIALIZER;;
```
Thank you! Fixed it in [5590cdd](https://github.com/CRaC/minicriu/pull/9/commits/5590cddb6f7d00caadbe7f8367612ee3fa63e08c)

i-pankrat commented on minicriu-client.c:
```diff
@@ -42,13 +42,55 @@
+	char filepath[MC_MAX_PHDRS][512];
+} nt_file;
+
+Elf64_Ehdr ehdr;
```
You are right, I don't have a reason either. [8d81a5d](https://github.com/CRaC/minicriu/pull/9/commits/8d81a5d4057e839be12adfc358ed504d93980569)

i-pankrat commented on minicriu-client.c:
```diff
@@ -102,12 +144,295 @@ static int writefile(const char *file, const char *buf, size_t len) {
+	int phnum = 0;
+
+	// Create Elf header
+	memcpy(ehdr.e_ident, ELFMAG, SELFMAG);
```
Yeah, it helps to avoid manual initialization with zeros. [4c267eb](https://github.com/CRaC/minicriu/pull/9/commits/4c267ebc0c199b1ce764ae843f7aedf5e52b2c2e)

i-pankrat commented on minicriu-client.c:
```diff
@@ -102,12 +144,295 @@ static int writefile(const char *file, const char *buf, size_t len) {
+				int writtenZeroes = fwrite(paddingData, sizeof(paddingData), n, coreFile);
+				if (writtenZeroes != n) {
+					perror("Failed replace map content with zeroes. Failed to create checkpoint.");
+					fclose(proc_maps);
```
Fixed in [80296e7](https://github.com/CRaC/minicriu/pull/9/commits/80296e7ae32a37b342ae052ab754f637469ceab7)

rvansa commented:
Looks like you've addressed some review notes; do you intend to work on this further or should this be considered for actual merging? The `[vvar]` IMO does not need to be handled in this PR, barrier initialization would be nice to have but I found that standard posix semaphores don't have the desired API.

i-pankrat commented:
> Looks like you've addressed some review notes; do you intend to work on this further or should this be considered for actual merging? The `[vvar]` IMO does not need to be handled in this PR, barrier initialization would be nice to have but I found that standard posix semaphores don't have the desired API.

I desire to keep working on minicriu, but after my last exam on June 29. I thought I would find time during exams to develop minicriu, but it is more challenging for me than I thought. If this pull request suits you in terms of quality, then we can merge it. In that case, I will do what we talked about in the new pull requests. Or I can continue development later in this pull request.

rvansa commented:
Superseded by #11 ; please either close this and integrate #11 or push the fixup here and close it.
